guides/inventing-service.md

```diff
@@ -0,0 +1,884 @@
+`Handler` that accepts some `Request` struct and asynchronously sends it to
+someone on the internet. Our `Timeout` wrapper is actually useful here as well.
+`JsonContentType` probably isn't since its not the clients job to set response
+headers.
+
+Since our `Handler` trait is useful for defining both servers and clients,
+`Handler` probably isn't an appropriate name anymore. A client doesn't handle an
+request, it sends it to a server and the server handles it. Lets instead call
+our trait `Service`:
+
+```rust
+trait Service<Request> {
+    type Response;
+    type Error;
+    type Future: Future<Output = Result<Self::Response, Self::Error>>;
+
+    fn call(&mut self, request: Request) -> Self::Future;
+}
+```
+
+This is actually _almost_ the `Service` trait defined in Tower. If you've been
+able to follow along until this point you now understand most of Tower. Besides
+the `Service` trait Tower also provides several utilities that implement
+`Service` by wrapping some other type that also implements `Service`. Exactly
+like we did with `Timeout` and `JsonContentType`. Things can be composed in ways
+similar to what we've just did.
+
+Some example services provided by Tower:
+
+- `Timeout` - This is pretty much identical to the timeout we have built.
+- `Retry` - To automatically retry failed requests.
+- `RateLimit` - To limit the number of requests a service can handle over a
+  period of time.
+
+Types like `Timeout` and `JsonContentType` are typically called "middleware"
+since they wrap another `Service` and transform the request or response in some
+way. Whereas types like `RequestHandler` is typically called "leaf services"
+since it sits at the leaves of a tree of nested services. The actual responses
+are normally produced in leaf services and modified by middlewares.
+
+The only thing left to talk about is "backpressure" and `poll_ready`.
+
+## Backpressure
+
+Imagine you wanted to write a rate limit middleware that wraps a `Service` and
+puts a limit on the maximum number of concurrent requests the underlying service
+will receive. This would be useful if you had some service that had a hard upper
+limit on the amount of load it could handle.
+
+With our current `Service` trait we don't really have a good way to implement
+something like that. We could try:
+
+```rust
+impl<R, T> Service<R> for ConcurrencyLimit<T> {
+    fn call(&mut self, request: R) -> Self::Future {
+        // 1. Check a counter for the number of running requests
+        //    currently.
+        // 2. If there is capacity left send the request to `T`
+        //    and increment the counter.
+        // 3. If not somehow wait until capacity becomes available.
+    }
+}
+```
+
+If there is no capacity left we have to wait and somehow get notified when
+capacity becomes available. So if the service is receiving a lot of requests we
+risk spending a lot of time waiting. It would be nice if we had a way to tell
```

Comment:
"spending a lot of time waiting" struck me as a somewhat odd way of expressing the problem. After all, the future will just be asleep, so what's the issue? You later get onto what I think is the _real_ problem, namely that you have to already have the request in memory in order to invoke `call`, and therefore you can only exert backpressure after you've already accepted a request, which leads to buffering up a lot of requests internally in the server. I think it's probably worth hoisting that point up here directly.

Reply:
Thats a very good point! I'll fix that :)